graphql_compiler/compiler/workarounds/orientdb_query_execution.py

```diff
@@ -0,0 +1,341 @@
+      or have a "where:" clause that references a non-local (i.e. tagged) field.
+    - Find all remaining query steps with "where:" clauses that reference only local fields.
+    - If any are found, we guide our actions from assumption 1 above:
+        - Ensure they have a defined "class:" -- i.e. the OrientDB scheduler will consider them
+          valid start points.
+        - Then, prune all other query steps (ones without such "where:" clauses) by removing their
+          "class:" clause, making them invalid as query start points for OrientDB's scheduler.
+    - If none are found, we guide our actions from assumption 2 above:
+        - Ensure that all query points not inside fold, optional, or recursion scope contain
+          a "class:" clause. That increases the number of available query start points,
+          so OrientDB can choose the start point of lowest cardinality.
+"""
+
+from ..blocks import CoerceType, QueryRoot, Recurse, Traverse
+from ..expressions import ContextField, ContextFieldExistence
+from ..ir_lowering_match.utils import convert_coerce_type_and_add_to_where_block
+from ..helpers import get_one_element_collection_value
+
+
+def _is_local_filter(filter_block):
+    """Return True if the Filter block references no non-local fields, and False otherwise."""
+    # We need the "result" value of this function to be mutated within the "visitor_fn".
+    # Since we support both Python 2 and Python 3, we can't use the "nonlocal" keyword here:
+    # https://www.python.org/dev/peps/pep-3104/
+    # Instead, we use a dict to store the value we need mutated, since the "visitor_fn"
+    # can mutate state in the parent scope, but not rebind variables in it without "nonlocal".
+    # TODO(predrag): Revisit this if we drop support for Python 2.
+    result = {
+        'is_local_filter': True
+    }
+    filter_predicate = filter_block.predicate
+
+    def visitor_fn(expression):
+        """Expression visitor function that looks for uses of non-local fields."""
+        non_local_expression_types = (ContextField, ContextFieldExistence)
+
+        if isinstance(expression, non_local_expression_types):
+            result['is_local_filter'] = False
+
+        # Don't change the expression.
+        return expression
+
+    filter_predicate.visit_and_update(visitor_fn)
+
+    return result['is_local_filter']
+
+
+def _classify_query_locations(match_query):
+    """Classify query locations into three groups: preferred, eligible, ineligible.
+
+    - Ineligible locations are ones that cannot be the starting point of query execution.
+    - Preferred locations are ones that are eligible to be the starting point, and also have
+      an associated "where:" clause that references no non-local fields -- only local fields,
+      literals, and variables.
+    - Eligible locations are all locations that do not fall into either of these two categories.
```

Comment:
The nomenclature here is a bit strange. Shouldn't eligible locations include preferred locations as well?

Reply:
They are, but they are "merely" eligible as opposed to being preferred. Naming is hard, so I tried to define each of these classes explicitly -- if you have a suggestion for a better name, I'm happy to add that.